### DISCUSSION TITLE
AttributeError: module 'pyswmm' has no attribute 'lib'

### DISCUSSION THREAD

**Discussion by yangqiu1018** [pyswmm/pyswmm#553] [Q&A] [answered]
Hi everyone, I downloaded a piece of code from github.com that was uploaded in 2019, but when I run it on my computer, it reports the following error: AttributeError: module 'pyswmm' has no attribute 'lib'. Queried pyswmm and found that it does not have the module lib. May I ask what is the reason for this?

**Comment by yangqiu1018** [comment]
Adding details about the original code that reported the error: code “pyswmm.lib.use('libswmm5_hs.so')” in a defined function

**Reply by karosc** [reply, marked as answer]
Your code might depend on pyswmm<=0.6.2 when [pyswmm had a lib module](https://github.com/pyswmm/pyswmm/blob/v0.6.2/pyswmm/lib/__init__.py)

**Reply by yangqiu1018** [reply]
Thank you very much for your answer!
May I ask what module/code can be used to realize this function in the latest version of pyswmm (without the module lib)? I would be very grateful for your answer!

**Reply by karosc** [reply]
You should just be able to remove that call to `pyswmm.lib.use`, unless the code you downloaded depends on a custom compilation of swmm. It's a bit more complex if that's the case and you might be unable to upgrade pyswmm....it all depends on the implementation of the shared object.

**Reply by bemcdonnell** [reply]
@yangqiu1018, it sounds like your project depended on a different version of SWMM. Just wanted to mention that if your archived project from 2019 depends on specific features compiled inside libswmm_hs.so, you might see different results from the engine behind pyswmm-2

**Reply by yangqiu1018** [reply]
> You should just be able to remove that call to `pyswmm.lib.use`, unless the code you downloaded depends on a custom compilation of swmm. It's a bit more complex if that's the case and you might be unable to upgrade pyswmm....it all depends on the implementation of the shared object.

Yes, I have to do a custom compilation of swmm for a job I'm working on. According to your suggestions, I removed this line of code, but there are other issues that haven't been resolved, and the whole program is currently not running, so not sure how much this line of code affects the whole program.
Could you tell me what the function of this module is if you know, and I'll try to see if there's any other way I can accomplish this!

**Reply by yangqiu1018** [reply]
> @yangqiu1018, it sounds like your project depended on a different version of SWMM. Just wanted to mention that if your archived project from 2019 depends on specific features compiled inside libswmm_hs.so, you might see different results from the engine behind pyswmm-2

Thanks for your mention!
I donnot know the function of lib module and the whole program isn't currently running, so I'm not sure how much this module affects the whole program.

**Reply by bemcdonnell** [reply]
@yangqiu1018, the lib module was used to enable you to use a different version of SWMM if you had different features / functions within it.  Do you have access to the c-code that they would have modified to produce your version of SWMM ? It might just need a basic migration in Python rather than the c code

**Reply by yangqiu1018** [reply]
@bemcdonnell ，Thank you so much for your answer!
Although I don't understand why this code, namely "pyswmm.lib.use('libswmm5_hs.so')", was added here at the moment,  I realized that the project just needs to call swmm multiple times for simulation and thus to optimize the action of the orifice control after understanding the function of this module, instead of modifying the original code of swmm. 
Next, I'll remove this line of code and see if it has any significant impact after modifying the other issues. 
Thank you again!

**Reply by bemcdonnell** [reply]
@yangqiu1018, feel free to post the relevant code if you need a second opinion on how it is implemented.
